File: bumperbot_ws/src/bumperbot_controller/bumperbot_controller/simple_controller.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import TwistStamped
import numpy as np

class SimpleController(Node):
  def __init__(self):
    super().__init__("simple_controller_py")

    # param
    self.declare_parameter("wheel_radius", 0.033) # 바퀴 반지름
    self.declare_parameter("wheel_separation", 0.17) # 바퀴 사이 거리

    self.wheel_radius = self.get_parameter("wheel_radius").get_parameter_value().double_value
    self.wheel_separation = self.get_parameter("wheel_separation").get_parameter_value().double_value

    self.get_logger().info(f"using wheel radius: {self.wheel_radius}")
    self.get_logger().info(f"using wheel separation: {self.wheel_separation}")

    self.wheel_cmd_pub = self.create_publisher(
      msg_type=Float64MultiArray, 
      topic="simple_velocity_controller/commands",
      qos_profile=10
    )

    self.vel_sub = self.create_subscription(
      msg_type=TwistStamped,
      topic="bumperbot_controller/cmd_vel",
      callback=self.velCallback,
      qos_profile=10
    )

    self.speed_conversion = np.array([
      [self.wheel_radius/2, self.wheel_radius/2],
      [self.wheel_radius/self.wheel_separation, - self.wheel_radius/self.wheel_separation]]
    )

    self.get_logger().info(f"conversion metrix : {self.speed_conversion}")

  def velCallback(self, msg: TwistStamped):
    # 로봇의 선 각 속도
    robot_speed = np.array([[msg.twist.linear.x],[msg.twist.angular.z]])
    wheel_speed = np.linalg.inv(self.speed_conversion) @ robot_speed

    wheel_speed_msg = Float64MultiArray()
    # 선속도, 각속도
    wheel_speed_msg.data = [wheel_speed[1, 0], wheel_speed[0, 0]]
    self.wheel_cmd_pub.publish(wheel_speed_msg)

def main(args=list[str] | None):
  rclpy.init(args)
  node = SimpleController()
  rclpy.spin(node)
  node.destroy_node()
  rclpy.shutdown()

if __name__ == "__main__":
  main()

File: bumperbot_ws/src/bumperbot_controller/src/simple_controller.cpp
#include "bumperbot_controller/simple_controller.hpp"
#include <Eigen/Geometry>

using std::placeholders::_1;

SimpleController::SimpleController(const std::string &name)
: Node(name) 
{
  declare_parameter("wheel_radius", 0.033);
  declare_parameter("wheel_separation", 0.17);

  wheel_radius = get_parameter("wheel_radius").as_double();
  wheel_separation = get_parameter("wheel_separation").as_double();

  RCLCPP_INFO_STREAM(get_logger(), "wheel radius: " << wheel_radius);
  RCLCPP_INFO_STREAM(get_logger(), "wheel separation: " << wheel_separation);

  wheel_cmd_pub = create_publisher<std_msgs::msg::Float64MultiArray>("/simple_velocity_controller/commands", 10);
  vel_sub = create_subscription<geometry_msgs::msg::TwistStamped>(
    "/bumperbot_controller/cmd_vel", 10, std::bind(&SimpleController::velCallback, this, _1)
  );
  speed_conversion << wheel_radius/2, wheel_radius/2, wheel_radius/wheel_separation, -wheel_radius/wheel_separation;

  RCLCPP_INFO_STREAM(get_logger(), "speed conversion matrix: \n" << speed_conversion);
}

void SimpleController::velCallback(const geometry_msgs::msg::TwistStamped &msg) {
  Eigen::Vector2d robot_speed(msg.twist.linear.x, msg.twist.angular.z);
  Eigen::Vector2d wheel_speed = speed_conversion.inverse() * robot_speed;
  std_msgs::msg::Float64MultiArray wheel_speed_msg;
  wheel_speed_msg.data.push_back(wheel_speed.coeff(1));
  wheel_speed_msg.data.push_back(wheel_speed.coeff(0));

  wheel_cmd_pub->publish(wheel_speed_msg);
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleController>("simple_controller");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}